### PR TITLE
Add utility functions to support graph analysis and viz

### PR DESCRIPTION
**Pull request by matthewmcneely**
This PR adds a utility function to extract nodes and edges from embedded JSON (usually from results of a DQL or GraphQL request), and adds some unit tests for the same. This tees up some Jupyter notebooks that I'd like to add to the examples folder once this PR is merged.

I also updated the acl_test to skip if the dgraph image is not found in the tester's PATH.

### CODE REVIEW

**Comment by skrdgraph** [comment]
👍 looks good

**Comment by matthewmcneely** [comment]
> 👍 looks good

Cool. Hopefully a reviewer (@joshua-goldstein :) can approve... We'll need to bump the version and publish to PyPI in order to be able to use the utility function in the Notebooks that I have planned for the `examples` folder.

**Comment by skrdgraph** [comment]
> ump the version and publish to PyPI in order to be able to use the utility function in the Notebooks that I have planned for the `examples` folder.

@matthewmcneely we are figuring out the publish step to pypi atm .. so getting this change as part of the next client release is better.

**Review by joshua-goldstein — APPROVED** [review]
Looks good :+1:   Thanks Matthew!

**Comment by mangalaman93** [comment]
I think instead of a special function, we could think of general API that can convert the returned JSON into more useful things. For example, it could be converted into nodes & edges, it could be converted into a Pandas dataframe and so on. We shouldn't expose small utility functions as part of the API for our clients.

**Review by mangalaman93 — DISMISSED** [review]
We also need to update the docs and add examples to show that this API is generic enough to be added to the client.

**Comment by joshua-goldstein** [comment]
To address @mangalaman93 concerns, we could add a new directory and name it something like `experimental` and put this function inside a module there (with the associated tests).  Everything in there will have the caveat that the API's can change at any time.  At the same time I think this utility function can be useful and should be accessible to anyone pulling pydgraph, so this would address both concerns.  @matthewmcneely @mangalaman93 thoughts?

**Comment by skrdgraph** [comment]
https://peps.python.org/pep-0702/#rationale .. hey @mangalaman93 I don't think this is a blocker. You guys will do a massive revamp of your client. 

You can add an @experimental annotation of some kind to indicate what you are saying.

**Comment by joshua-goldstein** [comment]
^ Nice, I didn't realize we could annotate functions like this in Python.  This looks good and I think is more straightforward to implement (adding an `:experimental:` annotation to the function) to stress that the function API might change in the future.

**Comment by mangalaman93** [comment]
@matthewmcneely  Additionally, I would also like to understand why we think this would be a useful function to our community. We should update the documentation as well in the same PR if we expect the community to use it. If we do not expect the community to use it, why are we considering adding this function to begin with.

**Comment by matthewmcneely** [comment]
> it could be converted into a Pandas dataframe and so on

I had thought of that, but wanted to keep external dependencies out of the package. Dicts of nodes and arrays of edges are the building blocks to Pandas or Arrow or Parquet, etc, etc.

> We shouldn't expose small utility functions as part of the API for our clients

Fair point. But the hard truth is this: for any graph analysis and visualization work, people will need to convert nested JSON to nodes and edges. My motivation is based around the idea that we should have a number of Notebooks in the examples folder that would help people better understand pydgraph (and Dgraph in general). We could replicate this utility code every time we publish a Notebook, but that seems anti-DRY.

**Comment by matthewmcneely** [comment]
> I would also like to understand why we think this would be a useful function to our community

As stated above, any graph analysis with say, igraph or networkx and visualization with say, graphistry or cytoscape, will inevitably involve converting nested json to dicts of nodes and arrays of edges. This code provides that utility.

**Comment by joshua-goldstein** [comment]
Thanks @matthewmcneely .  I agree this is a useful thing to have in the client and other people may find it useful.  We can move this function into a new module marked as experimental in some way and then merge this.

**Comment by matthewmcneely** [comment]
@joshua-goldstein OK, reworked the conversion code to a module named `convert.py`. Maybe later we could add other transformational funcs in there.

I made a note in the function comments that the func is experimental and may change in the future.

**Review by joshua-goldstein — APPROVED** [review]
Changes look good :+1:

File: tests/test_acl.py [hunk 1]
```diff
@@ -20,12 +20,13 @@
...
 
 import logging
-import os
+import shutil
 import unittest
 
 from . import helper
 import pydgraph
 
+@unittest.skipIf(shutil.which('dgraph') is None, 'Dgraph binary not found.')
```
Comment (mangalaman93) [inline tests/test_acl.py:29]: why are we adding this?

File: tests/test_util.py [hunk 2]
```diff
@@ -20,28 +20,24 @@
...
 
 from pydgraph import util
-import pydgraph
 
-from . import helper
-
-
-class TestIsString(unittest.TestCase):
-    """Tests is_string utility function."""
+class TestUtil(unittest.TestCase):
```
Comment (mangalaman93) [inline tests/test_util.py:24]: why did we change the function name here?

File: tests/test_convert.py [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+
+import unittest
+import json
+
+from pydgraph import convert
+
+class TestConvert(unittest.TestCase):
+    """Tests conversion functions."""
+
+    def test_extract_nodes_edges(self):
```
Comment (mangalaman93) [inline tests/test_convert.py:25]: Wouldn't it be better if we could use Dgraph responses and apply the extract_nodes function instead? What if the response changes tomorrow but we forget to improve/modify this function?

**Review by mangalaman93 — COMMENTED** [review]
I think this PR needs more work and docs. @matthewmcneely see if you can get that done in a separate PR. Thanks

File: tests/test_acl.py (hunk 1 above)
Comment (joshua-goldstein) [reply tests/test_acl.py:29]: this test requires the dgraph binary to be located on the client's path

File: tests/test_util.py (hunk 2 above)
Comment (joshua-goldstein) [reply tests/test_util.py:24]: the class should be named after the python module (i.e. file).  This is the pattern in the other test files.  It is clear when seen in context: https://github.com/dgraph-io/pydgraph/blob/866d3173bd1c556e819910c7017cd5dd80c5c04a/tests/test_util.py#L24-L27

File: tests/test_convert.py (hunk 3 above)
Comment (joshua-goldstein) [reply tests/test_convert.py:25]: If I understand you correctly, you are describing an integration test, whereas this is a simple unit test.  If the response changes tomorrow (unlikely, but maybe months later) we will see a failure here and fix it.  Writing an integration test here might be a good exercise for one of our junior engineers.  :smile_cat:

File: tests/test_acl.py (hunk 1 above)
Comment (mangalaman93) [reply tests/test_acl.py:29]: But this is the kinda thing that ends up in CI not running the tests at all. We should really consider undoing it.

File: tests/test_util.py (hunk 2 above)
Comment (mangalaman93) [reply tests/test_util.py:24]: We can change the name of the file. We should just avoid the word "util" everywhere we can. It has not value in its meaning.

File: tests/test_convert.py (hunk 3 above)
Comment (mangalaman93) [reply tests/test_convert.py:25]: Will you write one?

Comment (all-seeing-code) [reply tests/test_convert.py:25]: Open an issue and tackle it there in github? We can tag it as a `good-first-issue` or something similar. I don't see the reason to block this (well it's already merged so...). This is a nifty utility to integrate with graphistry etc.

Comment (joshua-goldstein) [reply tests/test_convert.py:25]: I can open an issue and think of who to give it to.  Apologies if that came off as a bit acerbic - I literally meant that it is a nice exercise and someone can learn something from doing it.

**Pull request merged**